### PR TITLE
helm: apply Role and RoleBinding for envoy gateway in namespaced mode

### LINKED ISSUE #1619: Apply EG Role and RoleBinding when running in namespaced mode
*Description*:
Envoy Gateway recently added support to allow users to limit the namespaces Envoy Gateway watches to gather resources in https://github.com/envoyproxy/gateway/issues/1117

This also means Envoy Gateway can use namespaced `Role` & `RoleBinding` resources instead of `ClusterRole` https://github.com/envoyproxy/gateway/blob/main/charts/gateway-helm/templates/generated/rbac/roles.yaml reducing the permissions needed to operate.

### PR DESCRIPTION
**Pull request by shawnh2**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1619

### CODE REVIEW

**Comment by arkodg** [comment]
thanks for working on this @shawnh2, sharing some thoughts, @Xunzhuo would also love your thoughts on this
* maybe its time to stop auto generating the `ClusterRole` https://github.com/envoyproxy/gateway/blob/2edcafe430a2ac1e61ccb4d1a0f5a5ce00ad5a08/tools/make/kube.mk#L34 and just manage it ourselves. This allows us to use your current logic to decide if we need to create a `ClusterRole` or a `Role`
* I think we need `ClusterRole` for `GatewayClass` even for the namespaced mode, so we can split up roles into 2 resources, one for the gateway class cluster role and the other for everything else (cluster role for regular mode and role for namespaced mode)
*  Can we use `{{ include "eg.fullname" $ }}-envoy-gateway-role` instead of `{{ include "eg.fullname" $ }}-{{ $ns }}-envoy-gateway-role` since they are created in different namespaces so the `ns` separation shouldnt matter ?
* Would be good to test this out with the steps highlighted in https://gateway.envoyproxy.io/v0.4.0/user/deployment-mode.html#multi-tenancy and see that it works

**Comment by shawnh2** [comment]
1. sure we can mange `ClusterRole` and `Role` ourselves
2. +1, we should create `ClusterRole` and `Role` resources no matter what, but create `RoleBinding` if we run in namespaced mode, otherwise create `ClusterRoleBinding`
3. thanks for pointing it out, missing that oops
4. thanks for heads up, i will test it out

File: charts/gateway-helm/templates/envoy-gateway-rbac.yaml
Comment (arkodg) [inline charts/gateway-helm/templates/envoy-gateway-rbac.yaml:1]: pretty close, this looks good 
what I was referring to earlier is you will need to create a `ClusterRole` + `ClusterRoleBinding` for a `GatewayClass` by default / always since GatewayClass is always cluster scoped 
https://gateway-api.sigs.k8s.io/api-types/gatewayclass/

Comment (arkodg) [reply charts/gateway-helm/templates/envoy-gateway-rbac.yaml:1]: do you see the same error if you replace `Role` with `ClusterRole` , but use `RoleBinding` ?

Comment (shawnh2) [reply charts/gateway-helm/templates/envoy-gateway-rbac.yaml:1]: yes, the error is same

Comment (arkodg) [reply charts/gateway-helm/templates/envoy-gateway-rbac.yaml:1]: hmm any ideas @envoyproxy/gateway-reviewers & @envoyproxy/gateway-maintainers ?

Comment (chauhanshubham) [reply charts/gateway-helm/templates/envoy-gateway-rbac.yaml:1]: just thinking out loud, whether it has something to do with the scope of the [manager](https://github.com/envoyproxy/gateway/blob/7c6aa9ea849667d40d0c74c2acf37c8664068329/internal/provider/kubernetes/kubernetes.go#L51). not sure though. The `at the cluster scope` pointed me in that direction.

Comment (shawnh2) [reply charts/gateway-helm/templates/envoy-gateway-rbac.yaml]: it has been tested out using https://gateway.envoyproxy.io/v0.4.0/user/deployment-mode.html#multi-tenancy, and it works great!

File: api/config/v1alpha1/envoygateway_types.go
Comment (arkodg) [inline api/config/v1alpha1/envoygateway_types.go:188]: is `omitempty` needed here ?

Comment (shawnh2) [reply api/config/v1alpha1/envoygateway_types.go]: of course

**Comment by shawnh2** [comment]
> LGTM thanks @shawnh2, since this PR makes an API change (and affects the docs dir), lets merge this after v0.5 releases. In the interim would be great if you can share the rbac o/p for default case as well as namespaces case in this PR, thanks !

sure, i added some usage in https://gateway.envoyproxy.io/latest/user/deployment-mode.html

File: docs/latest/user/deployment-mode.md
Comment (arkodg) [inline docs/latest/user/deployment-mode.md:34]: lets edit existing `helm` cmd instead of adding one

Comment (shawnh2) [reply docs/latest/user/deployment-mode.md]: updated

File: docs/latest/user/deployment-mode.md
Comment (shawnh2) [reply docs/latest/user/deployment-mode.md:31]: updated

**Review by arkodg — APPROVED** [review]
LGTM thanks ! lets wait for v0.5 to land before merging this one

**Review by zirain — APPROVED** [review]

**Comment by zirain** [comment]
@arkodg ready to merge？

**Pull request merged**
